AP_DDS: Local Pose Publisher

pedro-fuoco opened the pull request:
Resolves #23278, here is the PR working on SITL:
![Screenshot from 2023-04-14 00-08-55](https://user-images.githubusercontent.com/62964137/231931691-8fe8a6ec-f366-4c22-8ccd-e6c6eba1d385.png)
The NED to ENU conversion seemed fine when I analysed the position coordinates of a plane flying in SITL. As you can see in the screenshot the plane went North-West after takeoff which resulted in a negative X axis (points towards east in ENU) and a positive Y axis (points towards north in ENU). Z is positive after takeoff.

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -160,6 +163,31 @@ void AP_DDS_Client::populate_static_transforms(tf2_msgs_msg_TFMessage& msg)
+    Quaternion orientation;
+    if (ahrs.get_quaternion(orientation))
+    {
+        msg.pose.orientation.x = orientation[2];
```
In the static transforms for GPS PR, I linked the documentation to the conventions as a comment and showed how it mapped. Can you do the same here? That will make it easier to validate your math through inspection. We will also need to validate the orientation, perhaps through Foxglove 3D by checking bank angles.

pedro-fuoco replied:
Sure! Added the comments. Will work on validating the orientation

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -160,6 +163,31 @@ void AP_DDS_Client::populate_static_transforms(tf2_msgs_msg_TFMessage& msg)
+void AP_DDS_Client::update_topic(geometry_msgs_msg_PoseStamped& msg)
+{
+    update_topic(msg.header.stamp);
+    strcpy(msg.header.frame_id, ENU_FRAME_ID);
```
Frame name should be "base_link", which is already set as `BASE_LINK_FRAME_ID` 
https://automaticaddison.com/coordinate-frames-and-transforms-for-ros-based-mobile-robots/

pedro-fuoco replied:
Done! 
![Screenshot from 2023-04-14 01-03-29](https://user-images.githubusercontent.com/62964137/231939333-e2b31118-2293-460d-971f-61f40833abad.png)

Ryanf55 replied:
This section of the REP applies. 
```
The transform from odom to base_link is computed and broadcast by one of the odometry sources.

The transform from map to base_link is computed by a localization component. However, the localization component does not broadcast the transform from map to base_link. Instead, it first receives the transform from odom to base_link, and uses this information to broadcast the transform from map to odom.
```
https://answers.ros.org/question/50870/what-frame-is-sensor_msgsimuorientation-relative-to/
^ More insight on frames

https://answers.ros.org/question/210124/which-frame_id-do-i-usually-specify-for-the-navsatfix-header/
^ NavSatFix

Give me some time to put together a diagram and we can go over it.

pedro-fuoco replied:
> You should put in the frame_id in which the sensor is mounted (the transform from base_link).

I got it now, that makes sense

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -339,6 +367,21 @@ void AP_DDS_Client::write_static_transforms()
     }
 }
 
+void AP_DDS_Client::write_local_position_topic()
```
Please rename to `write_local_pose_topic` as it's a pose, not a position, since the pose encodes both position and orientation. Same name change applies elsewhere

pedro-fuoco replied:
Changed the PR name, the commit name and the variables.

pedro-fuoco commented:
![PoseOnFoxglove](https://user-images.githubusercontent.com/62964137/231943664-b2f00465-1e9d-4abd-a437-37515a80bd41.gif)
Orientations validated in foxglove!

Ryanf55 commented:
> ![PoseOnFoxglove](https://user-images.githubusercontent.com/62964137/231943664-b2f00465-1e9d-4abd-a437-37515a80bd41.gif) [ ![PoseOnFoxglove](https://user-images.githubusercontent.com/62964137/231943664-b2f00465-1e9d-4abd-a437-37515a80bd41.gif) ](https://user-images.githubusercontent.com/62964137/231943664-b2f00465-1e9d-4abd-a437-37515a80bd41.gif) [ ](https://user-images.githubusercontent.com/62964137/231943664-b2f00465-1e9d-4abd-a437-37515a80bd41.gif) @Ryanf55 Figured out how to validate orientations on foxglove, they are wrong apparently lol. Will fix them soon to follow the
> 
> ```
> // In ROS REP 103, it follows this convention
>         // X - Forward
>         // Y - Left
>         // Z - Up
>         // https://www.ros.org/reps/rep-0103.html#axis-orientation
> ```

In foxglove 3D, FYI I'm about 75% sure that X corresponds to red, so if the vehicle is traveling straight, red is pointing into the wind. Then green for left, and blue for up (towards sky). 

XYZ = RGB

It's hard to tell from the gif, but I'll verify this on my side when you think it's good. 

Glad we checked.

amilcarlucas commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -160,6 +162,43 @@ void AP_DDS_Client::populate_static_transforms(tf2_msgs_msg_TFMessage& msg)
+        // X - North
+        // Y - East
+        // Z - Down
+    // Furthermore, to follow ROS REP 103, it is necessary to switch X and Y, 
```
```suggestion
    // As a consequence, to follow ROS REP 103, it is necessary to switch X and Y, 
```

amilcarlucas commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -8,12 +8,14 @@
 
 static constexpr uint16_t DELAY_TIME_TOPIC_MS = 10;
 static constexpr uint16_t DELAY_NAV_SAT_FIX_TOPIC_MS = 1000;
+static constexpr uint16_t DELAY_LOCAL_POSE_TOPIC_MS = 33;
```
Is this a delay or is this a period??

pedro-fuoco replied:
Its the minimum delay necessary for another message to be posted. In the loop it works as a period, yes. This corresponds to publishing the message at a frequency of ~30Hz.
About the variable name I'm following the standard naming of the other topics, as can be seen on DELAY_NAV_SAT_FIX_TOPIC_MS and DELAY_TIME_TOPIC_MS

pedro-fuoco commented:
I researched quite a bit to see if I got the orientation right. 
[This is what I was looking for](https://github.com/mavlink/mavros/issues/49#issuecomment-51614130). 

We are not using a body-fixed frame on the local pose topic.
X-forward refers to a body-fixed frames, so there is no reason for X to be pointing forward in this case.

There should be a dicussion about using a body-fixed frame in the local velocity topic though

Ryanf55 commented:
Just tested in [ROS 2 humble NAV2 bringup](https://navigation.ros.org/getting_started/index.html#running-the-example). The frame shown in RVIZ below is the `base_link` frame. 
![image](https://user-images.githubusercontent.com/25047695/233913344-2ef27bc1-768a-4083-a484-f9ba22a6e0e2.png)

If `waffle` turns 45 degrees to the left, now facing northwest, it looks like this. I also zoomed out, so you can see `map` and `odom`, which are on top of each other. 
![image](https://user-images.githubusercontent.com/25047695/233913939-1919f829-2b97-4f93-9fd9-50b456f05e13.png)

`base_link` is in body frame, since it's attached to the robot body, so let's abide by the convention you posted above from TSC21's original clarifications, which also conveniently match the behavior in NAV2, which we hope to support easy integration with. 

Expectation (In SITL+Foxglove): 
If the drone is flying north, the red arrow (X) should point north, the green arrow (Y) points west, and the blue arrow (Z) points towards the sky. 
If the drone is flying northwest, the red arrow (X) should point northwest, the green arrow (Y) points southwest, and the blue arrow (Z) remains towards the sky. 

Hope that clears it up.

pedro-fuoco commented:
@Ryanf55 Thanks a lot for the clarifications! 
Just fixed it up to follow x-forward, y-left, z-up orientation! Here are my recent tests in foxglove:
![ezgif-5-25253dae3e](https://user-images.githubusercontent.com/62964137/234086613-629feb50-730f-4bb4-8122-712a0233d81a.gif)
Also updated the comments so that the transformations are clear.
